lib/tapioca/runtime/reflection.rb

```diff
@@ -174,14 +175,24 @@ def resolve_loc(locations)
         resolved_loc.absolute_path || ""
       end
 
-      sig { params(singleton_class: Module).returns(T.nilable(Module)) }
-      def attached_class_of(singleton_class)
-        # https://stackoverflow.com/a/36622320/98634
-        result = ObjectSpace.each_object(singleton_class).find do |klass|
-          singleton_class_of(T.cast(klass, Module)) == singleton_class
+      # On Ruby 3.2 or newer, Class defines an attached_object method that returns the
+      # attached class of a singleton class without iterating ObjectSpace. On older
+      # versions of Ruby, we fall back to iterating ObjectSpace.
+      if CLASS_ATTACHED_OBJECT
+        sig { params(singleton_class: Class).returns(T.nilable(Module)) }
+        def attached_class_of(singleton_class)
+          T.cast(singleton_class.attached_object, Module)
         end
+      else
+        sig { params(singleton_class: Class).returns(T.nilable(Module)) }
+        def attached_class_of(singleton_class)
+          # https://stackoverflow.com/a/36622320/98634
+          result = ObjectSpace.each_object(singleton_class).find do |klass|
+            singleton_class_of(T.cast(klass, Module)) == singleton_class
+          end
 
-        T.cast(result, Module)
+          T.cast(result, Module)
```

Comment:
This should be:
```suggestion
          T.cast(result, T.nilable(Module))
```
since there is a chance that `find` does not find a Module that has this as its singleton class.

I guess our previous code wasn't quite correct, it just worked since we were always giving it singleton classes of modules.